### PR TITLE
[Fix] Disabled form buttons while submitting

### LINKED ISSUE #7212: 🐛 Submit button not disabled during submission
> __Note__
This issue has been split off from  #7172

## 🐛 Bug

We have several places in the app where the submit button is not disabled while the request is in progress.  If the use clicks the submit button rapidly it is possible to submit duplicates.

## 🦋 Expected Behaviour

Submit button should be disabled during the request.

## 📋 Steps to Reproduce

It can be helpful to add a sleep statement to one of the Laravel middleware layers to simulate a slow response during testing.

1. Navigate to your profile fill in the form to add an experience
2. Rapidly click the submit button
3. Observe multiple duplicated responses

## ✅ Acceptance Criteria

- [ ] Submit buttons should not allow repeated form submission while a response is pending
- [ ] Check entire app for needed fixes

### PR DESCRIPTION
**Pull request by esizer**
🤖 Resolves #7212 

## 👋 Introduction

Updates our form submit buttons to be disabled while submitting a form preventing duplicate submissions.

## 🕵️ Details

This is partly to do with misunderstanding surrounding `react-hook-form` and some older implementations of `BasicForm`. It seems `react-hook-form` wants `methods.handleSubmit` to be passed a function that returns a `Promise`. Most of our forms actually await thee promise and return nothing. In order to fix this, I added the `fetching` state from `urql` to the disabled prop as well where needed.

I believe it would be in our interest to review forms and get some sort of standard. Our `BasicForm` is a good starting point but this specific issue stemmed from how we defined the `onSubmit` so it may be worth standardizing how we write the `onSubmit` so we can rely more on `react-hook-form` state.

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build `npm run dev`
2. Navigate to forms, confirming that clicking submit it rapid succession does not double submit
3. Make sure you specifically test the "Add experience" form since that was the form that prompted the ticket

### CODE REVIEW

File: apps/web/src/components/EmploymentEquity/dialogs/DialogFooter.tsx
Comment (petertgiles) [inline apps/web/src/components/EmploymentEquity/dialogs/DialogFooter.tsx:1]: Could you take a second look at the DEI dialogs?  I don't think this one is fixed yet.

https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/e6337259-474d-4c90-92eb-2dfd3963590f

File: apps/web/src/pages/Applications/ApplicationResumePage/ApplicationResumePage.tsx
Comment (petertgiles) [inline apps/web/src/pages/Applications/ApplicationResumePage/ApplicationResumePage.tsx:1]: Could you check the add experience page associated with this page?  I think it needs fixing still.

Uploading simplescreenrecorder-2023-07-25_10.46.09.mp4…

Comment (petertgiles) [reply apps/web/src/pages/Applications/ApplicationResumePage/ApplicationResumePage.tsx:1]: The "save and continue" on this page, too.

https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/46029584-1755-4999-a436-4f318196c496

File: apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx
Comment (petertgiles) [inline apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx:1]: Could you check the dialog on this page?

Uploading simplescreenrecorder-2023-07-25_10.53.41.mp4…

Comment (petertgiles) [reply apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx:1]: And the "save and continue"?

https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/b5073793-56a3-45ed-8406-2366a3435443

**Review by petertgiles — CHANGES_REQUESTED** [review]
Very cool!  Thanks for fixing these.  I found a few that need a second look, though.  I wouldn't be surprised if most of these are related.

**Comment by petertgiles** [comment]
> I believe it would be in our interest to review forms and get some sort of standard. Our BasicForm is a good starting point but this specific issue stemmed from how we defined the onSubmit so it may be worth standardizing how we write the onSubmit so we can rely more on react-hook-form state.

Good idea!  Any interest in contributing to the docs in this PR as well?  :sweat_smile:

File: apps/web/src/components/EmploymentEquity/dialogs/DialogFooter.tsx
Comment (esizer) [reply apps/web/src/components/EmploymentEquity/dialogs/DialogFooter.tsx:1]: Not sure if this was the best approach but I rolled all of them up into one commit.

[use urql to disable application submit buttons](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7318/commits/b3e2a6430f9f63e968f4e077da0159b01ef1e452)

File: apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx
Comment (petertgiles) [reply apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx:1]: The "save and continue" is fixed but the "link an experience" dialog is not fixed yet.

**Comment by vd1992** [comment]
Bouncing around, things all seem to work right, I can't replicate the issue with linking skills to experiences as it works for me. 

https://recordit.co/6sHhyRjp7N

Found other small issues though instead, spun off
#7390 
#7391

File: apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx
Comment (esizer) [reply apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx:1]: Updated the submission handler to be async and await the mutation. Seems to have fixed the issue.

[improve stability of skill form button disable](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7318/commits/7927ef1ff13c76bf1a5da8f47084fb409d7eb6ea)

Comment (petertgiles) [reply apps/web/src/pages/Applications/ApplicationSkillsPage/ApplicationSkillsPage.tsx:1]: Got it!  :dart:

**Review by petertgiles — APPROVED** [review]

**Pull request merged**